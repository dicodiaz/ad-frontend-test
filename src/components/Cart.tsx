'use client';

import { useCartContext } from '@/context';
import { CartItem } from '.';

export const Cart: React.FC = () => {
  const { cart } = useCartContext();

  return (
    <section className="flex flex-col gap-y-8 px-6 py-8">
      <div className="flex flex-col gap-y-3 text-[#3B3B3B]">
        <h1 className="text-2xl font-bold leading-7 tracking-wide">Your cart</h1>
        <span className="text-xl leading-6 tracking-wide">
          {cart.length} {cart.length === 1 ? 'item' : 'items'}
        </span>
      </div>
      <div className="flex flex-col">
        {cart.map((game) => (
          <CartItem key={game.id} game={game} />
        ))}
      </div>
    </section>
  );
};
