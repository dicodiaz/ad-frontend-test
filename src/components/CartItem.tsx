import { useCartContext } from '@/context';
import { Game } from '@/utils/endpoint';
import { HiX } from 'react-icons/hi';

export type CartItemProps = {
  game: Game;
};

export const CartItem: React.FC<CartItemProps> = ({ game }) => {
  const { id, image, genre, name, price, description, isNew } = game;
  const { dispatch } = useCartContext();

  const removeFromCart = () => {
    dispatch({ type: 'REMOVE_FROM_CART', payload: id });
  };

  return (
    <div className="flex flex-col gap-y-4 px-4 py-5 [&~&]:border-t-[0.5px] [&~&]:border-[#8F8F8F]">
      <div className="relative flex gap-x-3">
        {isNew && (
          <div className="absolute left-3 top-3 rounded bg-[#F5F5F4] px-3 py-2 leading-4 tracking-wide">
            New
          </div>
        )}
        <img src={image} alt={name} className="h-[135px] flex-1 object-cover" />
        <div>
          <button type="button" className="p-1" onClick={removeFromCart}>
            <HiX className="size-5 text-[#8F8F8F]" />
          </button>
        </div>
      </div>
      <div className="flex flex-col gap-y-11">
        <div className="flex flex-col gap-y-3">
          <span className="text-sm font-bold leading-4 text-[#737373]">{genre}</span>
          <div className="flex flex-col gap-y-2">
            <h4 className="text-lg font-bold leading-5 tracking-wide text-[#3B3B3B]">{name}</h4>
            <span className="leading-5 text-[#737373]">{description}</span>
          </div>
        </div>
        <div className="flex justify-end text-lg font-bold leading-5 tracking-wide text-[#3B3B3B]">
          {price.toLocaleString('en-US', { style: 'currency', currency: 'USD' })}
        </div>
      </div>
    </div>
  );
};
