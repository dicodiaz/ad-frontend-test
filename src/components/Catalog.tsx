import { getDataFromServer } from '@/services';
import { GameCard, GenreSelect, SeeMoreButton } from '.';

export type CatalogProps = {
  genre?: string;
  currentPage: number;
};

export const Catalog: React.FC<CatalogProps> = async ({ genre, currentPage }) => {
  const { games, availableFilters, totalPages } = await getDataFromServer({
    genre,
    page: currentPage,
  });

  return (
    <section className="flex flex-col gap-y-2">
      <div className="flex flex-col gap-y-8">
        <GenreSelect availableFilters={availableFilters} />
        <div className="h-px bg-[#EFEDF3]" />
        <div className="flex flex-col gap-y-6 px-6">
          {games.map((game) => (
            <GameCard key={game.id} game={game} />
          ))}
        </div>
      </div>
      {currentPage < totalPages && <SeeMoreButton currentPage={currentPage} />}
    </section>
  );
};
