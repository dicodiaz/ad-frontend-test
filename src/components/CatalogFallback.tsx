import { ImSpinner2 } from 'react-icons/im';

export const CatalogFallback: React.FC = () => {
  return (
    <div className="flex justify-center px-6">
      <ImSpinner2 className="size-6 animate-spin" />
    </div>
  );
};
