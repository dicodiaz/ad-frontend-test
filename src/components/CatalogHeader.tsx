'use client';

import { Data } from '@/services';
import { useRouter, useSearchParams } from 'next/navigation';

export type CatalogHeaderProps = {
  availableFilters: Data['availableFilters'];
};

export const CatalogHeader: React.FC<CatalogHeaderProps> = ({ availableFilters }) => {
  const router = useRouter();
  const searchParams = useSearchParams();

  const handleGenreChange: React.ChangeEventHandler<HTMLSelectElement> = (e) => {
    const genre = e.target.value;
    const params = new URLSearchParams(searchParams);
    if (genre === 'all') {
      params.delete('genre');
    } else {
      params.set('genre', genre);
    }
    router.push(`?${params.toString()}`);
  };

  return (
    <header className="flex flex-col gap-y-8 border-b border-[#EFEDF3] px-6 py-8">
      <h1 className="text-2xl font-bold leading-7 tracking-[0.4px] text-[#3B3B3B]">TOP SELLERS</h1>
      <label className="flex gap-x-6 py-4">
        <span>Genre</span>
        <div className="my-[1px] w-px bg-[#3B3B3B]" />
        <select className="flex-1" onChange={handleGenreChange}>
          <option value="all">All</option>
          {availableFilters
            .sort((a, b) => a.localeCompare(b))
            .map((genre, index) => (
              <option key={index} value={genre}>
                {genre}
              </option>
            ))}
        </select>
      </label>
    </header>
  );
};
