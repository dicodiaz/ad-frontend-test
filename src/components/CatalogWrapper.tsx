import { getDataFromServer } from '@/services';
import { Catalog } from '.';

export type CatalogWrapperProps = {
  genre?: string;
};

export const CatalogWrapper: React.FC<CatalogWrapperProps> = async ({ genre }) => {
  const data = await getDataFromServer({ genre });

  return <Catalog data={data} />;
};
