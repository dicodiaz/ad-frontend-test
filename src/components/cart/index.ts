export * from './Cart';
export * from './CartItem';
export * from './OrderSummary';
