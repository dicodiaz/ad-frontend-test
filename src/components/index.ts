export * from './cart';
export * from './checkout';
export * from './home';
export * from './layout';
