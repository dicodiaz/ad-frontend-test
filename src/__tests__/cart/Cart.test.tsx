import { Cart } from '@/components';
import { render } from '@testing-library/react';

describe('Cart', () => {
  it('renders Cart unchanged', () => {
    const { container } = render(<Cart />);
    expect(container).toMatchSnapshot();
  });
});
