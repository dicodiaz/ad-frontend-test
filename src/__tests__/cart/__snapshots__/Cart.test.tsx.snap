// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Cart renders Cart unchanged 1`] = `
<div>
  <section
    class="flex flex-col gap-y-8 px-6 py-8 md:gap-y-10 md:px-20 md:py-10 xl:gap-y-12 xl:px-32 xl:py-12"
  >
    <div
      class="text-primary flex flex-col gap-y-3"
    >
      <h1
        class="text-2xl font-bold leading-7 tracking-wide md:text-3xl xl:text-4xl"
      >
        Your cart
      </h1>
      <h2
        class="text-xl leading-6 tracking-wide md:text-2xl md:leading-7"
      >
        1
         
        item
      </h2>
    </div>
    <div
      class="flex flex-col gap-y-8 md:flex-row md:gap-x-10 xl:gap-x-20"
    >
      <div
        class="flex flex-col md:flex-grow"
      >
        <div
          class="[&~&]:border-stroke-secondary flex flex-col gap-y-4 px-4 py-5 md:flex-row md:gap-x-6 [&~&]:border-t-[0.5px]"
          data-testid="cart-item"
        >
          <div
            class="relative flex gap-x-3 md:w-2/5 md:flex-none"
          >
            <div
              class="text-primary absolute left-3 top-3 rounded bg-stone-100 px-3 py-2 leading-4 tracking-wide"
            >
              New
            </div>
            <img
              alt="Cyberpunk 2077"
              class="h-[136px] object-cover md:h-auto"
              data-nimg="1"
              decoding="async"
              height="240"
              loading="lazy"
              src="/_next/image?url=%2Fgame-images%2Fcyberpunk2077.jpeg&w=1920&q=75"
              srcset="/_next/image?url=%2Fgame-images%2Fcyberpunk2077.jpeg&w=750&q=75 1x, /_next/image?url=%2Fgame-images%2Fcyberpunk2077.jpeg&w=1920&q=75 2x"
              style="color: transparent;"
              width="654"
            />
            <div
              class="md:hidden"
            >
              <button
                aria-label="Remove, Cyberpunk 2077"
                class="p-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="size-5 text-[#8F8F8F]"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 20 20"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="flex flex-col gap-y-3 md:py-2"
          >
            <span
              class="text-sm font-bold leading-4 text-neutral-500"
            >
              Action
            </span>
            <div
              class="flex flex-col gap-y-2"
            >
              <h3
                class="text-primary text-lg font-bold leading-5 tracking-wide md:text-xl md:leading-6"
              >
                Cyberpunk 2077
              </h3>
              <span
                class="min-h-[60px] leading-5 text-neutral-500"
              >
                An open-world, action-adventure story set in Night City.
              </span>
            </div>
            <div
              class="text-primary flex justify-end text-lg font-bold leading-5 tracking-wide md:text-xl md:leading-6"
            >
              $59.99
            </div>
          </div>
          <div
            class="hidden md:block"
          >
            <button
              aria-label="Remove, Cyberpunk 2077"
              class="p-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="size-5 text-[#8F8F8F]"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 20 20"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="flex flex-col gap-y-10 md:w-2/5 md:flex-none"
      >
        <div
          class="border-stroke-secondary flex flex-col gap-y-6 rounded-lg border-[0.5px] px-4 py-6"
        >
          <div
            class="flex flex-col gap-y-3"
          >
            <h3
              class="text-primary text-xl font-bold leading-6 tracking-wide md:text-2xl md:leading-7"
            >
              Order Summary
            </h3>
            <span
              class="text-lg leading-6 tracking-wide"
            >
              1
               
              item
            </span>
          </div>
          <div
            class="flex flex-col gap-y-6 py-5"
          >
            <div
              class="flex flex-col gap-y-3"
            >
              <div
                class="text-primary flex items-center justify-between gap-x-3 text-lg leading-6 tracking-wide"
              >
                <span>
                  Cyberpunk 2077
                </span>
                <span>
                  $59.99
                </span>
              </div>
            </div>
            <div
              class="bg-stroke-secondary h-px opacity-50"
            />
            <div
              class="text-primary flex justify-between text-xl font-bold leading-6 tracking-wide"
            >
              <h4>
                Order total
              </h4>
              <span>
                $59.99
              </span>
            </div>
          </div>
        </div>
        <a
          href="/checkout"
        >
          <button
            class="cta-fill-primary"
            type="button"
          >
            Checkout
          </button>
        </a>
      </div>
    </div>
  </section>
</div>
`;
